### ISSUE TITLE
How to include the encoder + decoder in a simple c program

### ISSUE THREAD

**Issue by GatCode** [google/lyra#73]
Hello, I am trying to use _lyra_ in a simple c program (hello world) which is built with _cmake_. As you can see in the screenshot, I can build _lyra_ with _bazel_ (run `cmake .. && make && ./main.o` from the build dir) but now I am stuck on how I can actually include _lyra_ in the main.c file and how I can link the `encoder_main` and `decoder_main` binaies to my main.o file. Can someone help me out? Thanks!
![screen](https://user-images.githubusercontent.com/19287308/145434919-ce8075b1-3bfa-46ab-92df-03b845ef2768.png)

**Comment by aluebs** [comment]
First, is there a strong reason to use cmake? Cause it should be easier to build and include it in bazel, just cause Lyra uses bazel.

Second, I am not sure what you are trying to do, but you most likely want to include [lyra_encoder](https://github.com/google/lyra/blob/main/lyra_encoder.h) and [lyra_decoder](https://github.com/google/lyra/blob/main/lyra_decoder.h), since that is the actual codec API. You can see an example on how to include them in [encoder_main_lib](https://github.com/google/lyra/blob/main/encoder_main_lib.cc#L35) and [decoder_main_lib](https://github.com/google/lyra/blob/main/decoder_main_lib.cc#L36), since those are just example file demos. You can also directly include these last ones, if you are dealing with files. An example on how to do that can be seen in [encoder_main](https://github.com/google/lyra/blob/main/encoder_main.cc#L23) and [decoder_main](https://github.com/google/lyra/blob/main/decoder_main.cc#L23). But these last ones aren't meant to be be included, since they are just binaries, not libraries.

**Comment by GatCode** [comment]
Sorry for my vague description - yes, basically, I want to use lyra to encode an audio stream. This stream will be sampled by a _nrf5340_ chip and therefore I also need cmake because the _nrf toolchain_ or _Zephyr_ in general, rely on cmake. My rather stupid hello world example, therefore is somewhat tailored to this use case.

Nevertheless, the problem with my setup is that:
1.) I need to make VSCode aware that I use lyra (screenshot - right side)
2.) somehow link the _encoder_main_ and/or _decoder_main_ to my _main.o_ inside cmake.

Basically this is where I am really stuck now.

![Screenshot 2021-12-09 at 6 43 16 PM](https://user-images.githubusercontent.com/19287308/145448142-83527eee-b5af-4911-abad-ee59fd3ceffb.png)

**Comment by aluebs** [comment]
That makes a lot of sense, thank you for clarifying.
1.) While I'm no cmake expert, it seems like your linker isn't finding the lyra_encoder header, so you probably need to add the lyra directory explicitly.
2.) You probably don't want to link encoder_main or decoder_main, since they are binaries, but lyra_encoder and lyra_decoder instead.

**Comment by GatCode** [comment]
Aaah thank  you very much! Yes this makes sense. I'll look into it and find a way on how I can add the lyra directory to cmake but without compiling it when I exec make.
Nevertheless, one last question: since I need to somehow have a .so (to my understanding), is there a way in bazel that I can also create a _lyra_encoder.so_ inside the lyra bin dir?

**Comment by aluebs** [comment]
Absolutely! You can build Lyra .so libraries using bazel:

```
bazel build -c opt :encoder_main
bazel build -c opt :decoder_main 
```

You can also add `--config=android_arm64` flag to build for Android ARM64 devices.

**Comment by GatCode** [comment]
Aah perfect, thank you very much!
